### DISCUSSION TITLE
Track nullability of variables based on a boolean variables when a NotNullWhen attribute is included

### DISCUSSION THREAD

**Discussion by espartia** [dotnet/csharplang#6085] [Language Ideas]
## Proposal
Allow associating and tracking of boolean values with a nullable variable to better check whether they could be null
This is specifically referring to a nullable variable modified or created by a function with a NotNullWhen attribute.
This means that it should be able to tell that an out or ref parameter will not be null when:
 - A function is used that has an out or ref parameter that has a NotNullWhen attribute
 - That function returns the value within the NotNullWhen attribute
 - That return value is stored in a variable
 - An if statement is used to check whether or not that value could be null based on the boolean
When assigned after running the function, the variable should no longer be associated with the nullability of the variable.

## Examples
These functions are used in the examples:
1. A function that outputs a string that is not null when the function returns true:
```cs
bool Example([NotNullWhen(true)] out string? output, bool WantNull = false) {
  if (WantNull) {
    output = null;
    return false;
  } else {
    output = "Hello, world!";
    return true;
  }
}
```
2. A function that requires a non-null string value:
```cs
void TestString(string output) {
  Console.WriteLine(output);
}
```
---
Currently seen as safe:
```cs
if (Example(out var output)) {
  TestString(output);
}
```
---
Code that is safe, but in C# will throw a warning:
```cs
bool IsNotNull = Example(out var output);
if (IsNotNull) {
  TestString(output);
}
```
---
Code that is not safe and should throw a warning:
```cs
bool IsNotNull = Example(out var output, true);
IsNotNull = true;
if (IsNotNull) {
  TestString(output);
}
```

## Use case
Checking whether a variable can be null based on the return value of a function with the NotNullWhen attribute specified on a ref or out parameter
 - This allows initializing and setting one or more values with a function that returns true if it succeeded, false if it didn't and the values are null, without it going out of scope when using an if statement that returns from a function or throws an exception.

## Example of what shouldn't produce a warning as a result of this
[This repository](https://github.com/catmeow72/notnullwhen-example) contains examples of what should and shouldn't produce a warning if this is implemented.

**Comment by CyrusNajmabadi** [comment]
@catmeow72 can you please provide some examples of what you mean?  Thanks!

**Reply by espartia** [reply]
Done.
